File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-0.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        result = []
        digitmap = {
            "2":["a","b","c"]
            ,"3":["d","e","f"]
            ,"4":["g","h","i"]
            ,"5":["j","k","l"]
            ,"6":["m","n","o"]
            ,"7":["p","q","r","s"]
            ,"8":["t","u","v"]
            ,"9":["w","x","y","z"]}
        
        def backtrack(i, curr):
            if i >= len(digits):
                if curr:
                    result.append(curr)
                return
            
            digitChars = digitmap[digits[i]]
            for letter in digitChars:
                curr += letter
                backtrack(i+1, curr)
                curr = curr[:-1]
        
        backtrack(0, "")

        return result
            

File: Data Structures & Algorithms/combinations/submission-0.java
class Solution {

    public List<List<Integer>> result;

    public List<List<Integer>> combine(int n, int k) {
        result = new ArrayList<>();
        backtrack(1, n, k, new ArrayList<>());
        return result;
    }

    public void backtrack(int first, int n, int k, List<Integer>curr){
        if (curr.size() == k){
            result.add(new ArrayList<>(curr));
            return;
        }

        for (int i = first; i <= n; i++){
            curr.add(i);
            backtrack(i+1, n, k, curr);
            curr.remove(curr.size()-1);
        }
    }
}